python/GafferSceneUITest/SceneGadgetTest.py

```diff
@@ -675,6 +691,76 @@ def testSetRenderer( self ) :
 			sg.setRenderer( "OpenGL" )
 			self.waitForRender( sg )
 
+	def __rayCircleIntersection( self, origin, direction, center, radius ) :
+
+		origin -= center
+		b = 2 * origin.dot( direction )
+		c = origin.dot( origin ) - radius * radius
+
+		discriminant = b * b - 4.0 * c
+		if discriminant < 0 :
+			return None
+
+		t = ( -b - math.sqrt( discriminant ) ) * 0.5
+
+		return origin + direction * t
+
+	@unittest.skipIf(
+		os.environ.get( "GAFFER_BUILD_ENVIRONMENT", "" ) == "gcc9",
+		"The gcc9 container does not support floating point depth buffers."
+	)
+	def testNormalAt( self ) :
+
+		s = Gaffer.ScriptNode()
+		s["s"] = GafferScene.Sphere()
+
+		sg = GafferSceneUI.SceneGadget()
+		sg.setRenderer( self.renderer )
+		sg.setScene( s["s"]["out"] )
+		sg.setMinimumExpansionDepth( 100 )
+
+		with GafferUI.Window() as w :
+			gw = GafferUI.GadgetWidget( sg )
+
+		w.setVisible( True )
+		self.waitForIdle( 1000 )
+
+		vp = gw.getViewportGadget()
+
+		vp.setPlanarMovement( False )
+		vpSize = imath.V2i( 500, 500 )
+		vp.setViewport( vpSize )
+
+		c = IECoreScene.Camera()
+		c.setProjection( "perspective" )
+		c.setFocalLength( 36 )  # Use 36 to make the view plane at Z = 0 2.0 units wide
+		c.setAperture( imath.V2f( 36, 36 ) )
+		vp.setCamera( c )
+
+		cameraTransform = imath.M44f()
+		cameraOrigin = imath.V3f( 0.0, 0.0, 2.0 )
+		cameraTransform.translate( imath.V3f( 0, 0, 2  ) )
+		vp.setCameraTransform( cameraTransform )
+
+		self.waitForRender( sg )
+
+		random.seed( 1 )
+
+		for i in range( 100 ) :
+			# `rasterPosition` from the top-left corner of the viewport, +Y going down
+			rasterPosition = imath.V2f( int( random.random() * vpSize.x ), int( random.random() * vpSize.y ) )
+			gadgetLineCenter = vp.rasterToGadgetSpace( rasterPosition + imath.V2f( 0.5 ), sg )
+
+			# Calculate the eye ray through `rasterPosition`, flipping the Y coordinate
+			# so +Y is going up. We use this instead of `gadgetLineCenter` to avoid the
+			# precision issues `rasterToGadgetSpace()` introduces.
+			vpPosition = ( 2.0 * imath.V2f( rasterPosition.x + 0.5, 500.0 - ( rasterPosition.y + 0.5 ) ) ) / imath.V2f( vpSize ) - imath.V2f( 1.0 )
+			rayDirection = ( imath.V3f( vpPosition.x, vpPosition.y, 0.0 ) - cameraOrigin ).normalized()
```

Comment:
Can you clarify what precision issues `rasterToGadgetSpace()` is introducing? I has assumed the precisions problems were usually from the reverse - `gadgetToRasterSpace()`. Minimally, I'd like to come out of this PR with a clear understanding of what future improvements we'd like to make, which I thought were primarily about handling the `0.5` offset more centrally.

Reply:
Even if `rasterToGadgetSpace()` was imprecise, I guess I also don't understand how comparing those results with a more accurate formulation would help.

Reply:
Thanks for nudging on that one, it turns out using `gadgetLineCenter` in `__rayCircleIntersection` works after all. I was probably trying to fix a deeper problem than that precision, fixed whatever that problem was then didn't try reverting back the custom camera ray calculation. I'll have a fixup for that shortly.

Reply:
Switched to using `gadgetLineCenter` in https://github.com/GafferHQ/gaffer/pull/5630/commits/6a5eb3719f25b0ca6f0fa65d5f9df83358ef2f05